File: midonet-api/src/main/java/org/midonet/api/filter/Condition.java
/*
 * Copyright (c) 2013-2014 Midokura SARL, All Rights Reserved.
 */
package org.midonet.api.filter;

import java.util.*;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.Pattern;

import org.midonet.api.UriResource;
import org.midonet.api.rest_api.BadRequestHttpException;
import org.midonet.api.validation.MessageProperty;
import org.midonet.midolman.rules.FragmentPolicy;
import org.midonet.odp.flows.IPFragmentType;
import org.midonet.packets.IPv4Addr;
import org.midonet.packets.IPv4;
import org.midonet.packets.IPv4Subnet;
import org.midonet.packets.MAC;
import org.midonet.util.Range;
import org.midonet.util.version.Since;

import static org.midonet.packets.Unsigned.unsign;
import static org.midonet.midolman.rules.Condition.NO_MASK;

/* Trace class */
public abstract class Condition extends UriResource {
    @Pattern(regexp = IPv4.regex,
             message = "is an invalid IP format")
    private String nwDstAddress;

    @Min(0)
    @Max(32)
    private int nwDstLength;

    @Pattern(regexp = IPv4.regex,
             message = "is an invalid IP format")
    private String nwSrcAddress;

    @Min(0)
    @Max(32)
    private int nwSrcLength;

    private boolean condInvert;
    private boolean matchForwardFlow;
    private boolean matchReturnFlow;
    private UUID[] inPorts;
    private boolean invInPorts;
    private UUID[] outPorts;
    private boolean invOutPorts;
    private UUID portGroup;
    private boolean invPortGroup;
    private UUID ipAddrGroupSrc;
    private boolean invIpAddrGroupSrc;
    private UUID ipAddrGroupDst;
    private boolean invIpAddrGroupDst;

    @Min(0x0600)
    @Max(0xFFFF)
    private Integer dlType;
    private boolean invDlType;
    private String dlSrc;
    @Since("2")
    private String dlSrcMask;
    private boolean invDlSrc;
    private String dlDst;
    @Since("2")
    private String dlDstMask;
    private boolean invDlDst;
    private int nwTos;
    private boolean invNwTos;
    private int nwProto;
    private boolean invNwProto;
    private boolean invNwSrc;
    private boolean invNwDst;

    @Since("2")
    @Pattern(regexp = FragmentPolicy.pattern,
             message = MessageProperty.FRAG_POLICY_UNDEFINED)
    private String fragmentPolicy;

    private Range<Integer> tpSrc;
    private Range<Integer> tpDst;

    private boolean invTpSrc;
    private boolean invTpDst;

    public boolean isCondInvert() {
        return condInvert;
    }

    public void setCondInvert(boolean condInvert) {
        this.condInvert = condInvert;
    }

    public boolean isMatchForwardFlow() {
        return matchForwardFlow;
    }

    public void setMatchForwardFlow(boolean matchForwardFlow) {
        this.matchForwardFlow = matchForwardFlow;
    }

    public boolean isMatchReturnFlow() {
        return matchReturnFlow;
    }

    public void setMatchReturnFlow(boolean matchReturnFlow) {
        this.matchReturnFlow = matchReturnFlow;
    }

    public UUID[] getInPorts() {
        return inPorts;
    }

    public void setInPorts(UUID[] inPorts) {
        this.inPorts = inPorts;
    }

    public boolean isInvInPorts() {
        return invInPorts;
    }

    public void setInvInPorts(boolean invInPorts) {
        this.invInPorts = invInPorts;
    }

    public UUID[] getOutPorts() {
        return outPorts;
    }

    public void setOutPorts(UUID[] outPorts) {
        this.outPorts = outPorts;
    }

    public boolean isInvOutPorts() {
        return invOutPorts;
    }

    public void setInvOutPorts(boolean invOutPorts) {
        this.invOutPorts = invOutPorts;
    }

    public boolean isInvPortGroup() {
        return invPortGroup;
    }

    public void setInvPortGroup(boolean invPortGroup) {
        this.invPortGroup = invPortGroup;
    }

    public UUID getPortGroup() {
        return portGroup;
    }

    public void setPortGroup(UUID portGroup) {
        this.portGroup = portGroup;
    }

    public UUID getIpAddrGroupSrc() {
        return ipAddrGroupSrc;
    }

    public void setIpAddrGroupSrc(UUID ipAddrGroupSrc) {
        this.ipAddrGroupSrc = ipAddrGroupSrc;
    }

    public boolean isInvIpAddrGroupSrc() {
        return invIpAddrGroupSrc;
    }

    public void setInvIpAddrGroupSrc(boolean invIpAddrGroupSrc) {
        this.invIpAddrGroupSrc = invIpAddrGroupSrc;
    }

    public UUID getIpAddrGroupDst() {
        return ipAddrGroupDst;
    }

    public void setIpAddrGroupDst(UUID ipAddrGroupDst) {
        this.ipAddrGroupDst = ipAddrGroupDst;
    }

    public boolean isInvIpAddrGroupDst() {
        return invIpAddrGroupDst;
    }

    public void setInvIpAddrGroupDst(boolean invIpAddrGroupDst) {
        this.invIpAddrGroupDst = invIpAddrGroupDst;
    }

    @Since("2")
    public String getFragmentPolicy() {
        return fragmentPolicy;
    }

    @Since("2")
    public void setFragmentPolicy(String fragmentPolicy) {
        this.fragmentPolicy = fragmentPolicy;
    }

    /**
     * Get the Data Layer Destination that this condition matches on
     *
     * @return A MAC address specified as "aa:bb:cc:dd:ee:ff"
     */
    public String getDlDst() {
        return dlDst;
    }

    /**
     * Gets the Data Layer Destination mask.
     *
     * @return
     *     Data layer destination mask in the form "ffff.0000.0000"
     */
    @Since("2")
    public String getDlDstMask() {
        return dlDstMask;
    }

    /**
     * Set the Data Layer Destination that this condition matches on
     *
     * @param dlDst
     *            A MAC address specified as "aa:bb:cc:dd:ee:ff"
     */
    public void setDlDst(String dlDst) {
        this.dlDst = dlDst;
    }

    /**
     * Set the Data Layer Destination mask. Bits that are set to zero
     * in the mask will be ignored when comparing a packet's source
     * MAC address against the condition's data layer source address.
     *
     * @param dlDstMask
     *     Mask in the form "ffff.0000.0000" (ignores the lower 32 bits).
     */
    @Since("2")
    public void setDlDstMask(String dlDstMask) {
        this.dlDstMask = dlDstMask;
    }

    /**
     * Set whether the match on the data layer destination should be inverted.
     * This will be stored but ignored until the DlDst has been set.
     *
     * @param invDlDst
     *            True if the condition should match packets whose data layer
     *            destination is NOT equal to the MAC set by 'setDlDst'
     *            False if the condition should match packets whose DlDst IS
     *            equal to that MAC
     */
    public void setInvDlDst(boolean invDlDst) {
        this.invDlDst = invDlDst;
    }

    /**
     * Find out whether this condition's match on the Data Layer Destination is
     * inverted.
     *
     * @return True if the condition matches packets whose data layer
     *         destination is NOT equal to the MAC set by 'setDlDst'. False if
     *         the condition matches packets whose DlDst is equal to that MAC.
     */
    public boolean isInvDlDst() {
        return invDlDst;
    }

    /**
     * Get the Data Layer Source that this condition matches on.
     *
     * @return A MAC address specified as "aa:bb:cc:dd:ee:ff"
     */
    public String getDlSrc() {
        return dlSrc;
    }

    /**
     * Gets the Data Layer Source mask.
     *
     * @return
     *     Data layer source mask in the form "ffff.0000.0000"
     */
    public String getDlSrcMask() {
        return dlSrcMask;
    }

    /**
     * Set the Data Layer Source address that this condition matches on
     *
     * @param dlSrc
     *            A MAC address specified as "aa:bb:cc:dd:ee:ff"
     */
    public void setDlSrc(String dlSrc) {
        this.dlSrc = dlSrc;
    }

    /**
     * Set the Data Layer Source mask. Bits that are set to zero in
     * the mask will be ignored when comparing a packet's source MAC
     * address against the condition's data layer source address.
     *
     * @param dlSrcMask
     *     Mask in the form "ffff.0000.0000" (ignores the lower 32 bits).
     */
    public void setDlSrcMask(String dlSrcMask) {
        this.dlSrcMask = dlSrcMask;
    }

    /**
     * Set whether the match on the data layer source should be inverted. This
     * will be stored but ignored until the DlSrc has been set.
     *
     * @param invDlSrc
     *            True if the condition should match packets whose data layer
     *            source is NOT equal to the MAC set by 'setDlSrc'. False if the
     *            condition should match packets whose DlSrc IS equal to that
     *            MAC
     */
    public void setInvDlSrc(boolean invDlSrc) {
        this.invDlSrc = invDlSrc;
    }

    /**
     * Find out whether this condition's match on the Data Layer Source is
     * inverted.
     *
     * @return True if the condition matches packets whose data layer source is
     *         NOT equal to the MAC set by 'setDlSrc'. False if the condition
     *         matches packets whose DlSrc is equal to that MAC.
     */
    public boolean isInvDlSrc() {
        return invDlSrc;
    }

    /**
     * Set the Data Layer Type (Ethertype) of packets matched by this condition
     *
     * @param dlType
     *            Ethertype value. We do not check the validity of the value
     *            provided: i.e. whether it's in the correct range for
     *            Ethertypes.
     */
    public void setDlType(Integer dlType) {
        this.dlType = dlType;
    }

    /**
     * Get the Data Layer Type (Ethertype) of packets matched by this condition
     *
     * @return The value of the Ethertype as a Integer if the condition matches
     *         Ethertype, otherwise null.
     */
    public Integer getDlType() {
        return dlType;
    }

    /**
     * Set whether the match on the data layer type should be inverted. This
     * will be stored but ignored until the DlType has been set.
     *
     * @param invDlType
     *            True if the condition should match packets whose data layer
     *            type is NOT equal to the Ethertype set by 'setDlType'. False
     *            if the condition should match packets whose DlType IS equal
     *            to that Ethertype.
     */
    public void setInvDlType(boolean invDlType) {
        this.invDlType = invDlType;
    }

    /**
     * Find out whether this condition's match on the Data Layer Type is
     * inverted
     *
     * @return True if the condition matches packets whose data layer type is
     *         NOT equal to the Ethertype set by 'setDlType'. False if the
     *         condition matches packets whose DlSrc is equal to that Ethertype
     */
    public boolean isInvDlType() {
        return invDlType;
    }

    public int getNwTos() {
        return nwTos;
    }

    public void setNwTos(int nwTos) {
        this.nwTos = nwTos;
    }

    public boolean isInvNwTos() {
        return invNwTos;
    }

    /**
     * @param invNwTos
     *            the invNwTos to set
     */
    public void setInvNwTos(boolean invNwTos) {
        this.invNwTos = invNwTos;
    }

    public int getNwProto() {
        return nwProto;
    }

    /**
     * @param nwProto
     *            the nwProto to set
     */
    public void setNwProto(int nwProto) {
        this.nwProto = nwProto;
    }

    public boolean isInvNwProto() {
        return invNwProto;
    }

    /**
     * @param invNwProto
     *            the invNwProto to set
     */
    public void setInvNwProto(boolean invNwProto) {
        this.invNwProto = invNwProto;
    }

    public String getNwSrcAddress() {
        return nwSrcAddress;
    }

    /**
     * @param nwSrcAddress
     *            the nwSrcAddress to set
     */
    public void setNwSrcAddress(String nwSrcAddress) {
        this.nwSrcAddress = nwSrcAddress;
    }

    public int getNwSrcLength() {
        return nwSrcLength;
    }

    /**
     * @param nwSrcLength
     *            the nwSrcLength to set
     */
    public void setNwSrcLength(int nwSrcLength) {
        this.nwSrcLength = nwSrcLength;
    }

    public boolean isInvNwSrc() {
        return invNwSrc;
    }

    /**
     * @param invNwSrc
     *            the invNwSrc to set
     */
    public void setInvNwSrc(boolean invNwSrc) {
        this.invNwSrc = invNwSrc;
    }

    public String getNwDstAddress() {
        return nwDstAddress;
    }

    /**
     * @param nwDstAddress the nwDstAddress to set
     */
    public void setNwDstAddress(String nwDstAddress) {
        this.nwDstAddress = nwDstAddress;
    }

    public int getNwDstLength() {
        return nwDstLength;
    }

    /**
     * @param nwDstLength the nwDstLength to set
     */
    public void setNwDstLength(int nwDstLength) {
        this.nwDstLength = nwDstLength;
    }

    public boolean isInvNwDst() {
        return invNwDst;
    }

    /**
     * @param invNwDst the invNwDst to set
     */
    public void setInvNwDst(boolean invNwDst) {
        this.invNwDst = invNwDst;
    }

    /**
     * @return the tpSrc
     */
    public Range<Integer> getTpSrc() {
        return tpSrc;
    }

    /**
     * @param tpSrc the tpSrc to set
     */
    public void setTpSrcStart(Range<Integer> tpSrc) {
        this.tpSrc = tpSrc;
    }

    /**
     * @return the invTpSrc
     */
    public boolean isInvTpSrc() {
        return invTpSrc;
    }

    /**
     * @param invTpSrc the invTpSrc to set
     */
    public void setInvTpSrc(boolean invTpSrc) {
        this.invTpSrc = invTpSrc;
    }

    /**
     * @return the tpDst
     */
    public Range<Integer> getTpDst() {
        return tpDst;
    }

    /**
     * @param tpDst the tpDst to set
     */
    public void setTpDst(Range<Integer> tpDst) {
        this.tpDst = tpDst;
    }

    /**
     * @param tpSrc the tpSrc to set
     */
    public void setTpSrc(Range<Integer> tpSrc) {
        this.tpSrc = tpSrc;
    }

    /**
     * @return the invTpDst
     */
    public boolean isInvTpDst() {
        return invTpDst;
    }

    /**
     * @param invTpDst
     *            the invTpDst to set
     */
    public void setInvTpDst(boolean invTpDst) {
        this.invTpDst = invTpDst;
    }

    protected org.midonet.midolman.rules.Condition makeCondition() {
        org.midonet.midolman.rules.Condition c =
            new org.midonet.midolman.rules.Condition();
        c.conjunctionInv = this.isCondInvert();
        c.matchForwardFlow = this.matchForwardFlow;
        c.matchReturnFlow = this.matchReturnFlow;
        if (this.getInPorts() != null) {
            c.inPortIds = new HashSet<UUID>(Arrays.asList(this.getInPorts()));
        } else {
            c.inPortIds = new HashSet<UUID>();
        }
        c.inPortInv = this.isInvInPorts();
        if (dlType != null)
            c.etherType = this.dlType;
        c.invDlType = this.isInvDlType();
        if (this.dlSrc != null)
            c.ethSrc = MAC.fromString(this.dlSrc);
        if (this.dlSrcMask != null)
            c.ethSrcMask = MAC.parseMask(this.dlSrcMask);
        c.invDlSrc = this.invDlSrc;
        if (this.dlDst != null)
            c.ethDst = MAC.fromString(this.dlDst);
        if (this.dlDstMask != null)
            c.dlDstMask = MAC.parseMask(this.dlDstMask);
        c.invDlDst = this.invDlDst;
        c.nwDstInv = this.isInvNwDst();
        if (this.getNwDstAddress() != null) {
            c.nwDstIp = new IPv4Subnet(
                IPv4Addr.fromString(this.getNwDstAddress()),
                this.getNwDstLength());
        }
        if (nwProto != 0)
            c.nwProto = (byte)nwProto;
        c.nwProtoInv = this.isInvNwProto();
        c.nwSrcInv = this.isInvNwSrc();
        if (this.getNwSrcAddress() != null) {
            c.nwSrcIp = new IPv4Subnet(
                IPv4Addr.fromString(this.getNwSrcAddress()),
                this.getNwSrcLength());
        }
        if (nwTos != 0)
            c.nwTos = (byte) nwTos;
        c.nwTosInv = this.isInvNwTos();
        c.fragmentPolicy = getAndValidateFragmentPolicy();
        if (this.getOutPorts() != null) {
            c.outPortIds = new HashSet<UUID>(Arrays.asList(this.getOutPorts()));
        } else {
            c.outPortIds = new HashSet<UUID>();
        }
        c.outPortInv = this.isInvOutPorts();
        if (tpDst != null) {
            c.tpDst = tpDst;
        }
        c.tpDstInv = this.isInvTpDst();
        if (tpSrc != null) {
            c.tpSrc = tpSrc;
        }
        c.tpSrcInv = this.isInvTpSrc();
        c.portGroup = this.portGroup;
        c.invPortGroup = this.invPortGroup;
        c.ipAddrGroupIdDst = this.ipAddrGroupDst;
        c.invIpAddrGroupIdDst = this.invIpAddrGroupDst;
        c.ipAddrGroupIdSrc = this.ipAddrGroupSrc;
        c.invIpAddrGroupIdSrc = this.invIpAddrGroupSrc;

        return c;
    }

    protected void setFromCondition(org.midonet.midolman.rules.Condition c) {
        this.setCondInvert(c.conjunctionInv);
        this.setInvInPorts(c.inPortInv);
        this.setInvOutPorts(c.outPortInv);
        this.setInvPortGroup(c.invPortGroup);
        this.setInvIpAddrGroupDst(c.invIpAddrGroupIdDst);
        this.setInvIpAddrGroupSrc(c.invIpAddrGroupIdSrc);
        this.setInvDlType(c.invDlType);
        this.setInvDlSrc(c.invDlSrc);
        this.setInvDlDst(c.invDlDst);
        this.setInvNwDst(c.nwDstInv);
        this.setInvNwProto(c.nwProtoInv);
        this.setInvNwSrc(c.nwSrcInv);
        this.setInvNwTos(c.nwTosInv);
        this.setInvTpDst(c.tpDstInv);
        this.setInvTpSrc(c.tpSrcInv);

        this.setMatchForwardFlow(c.matchForwardFlow);
        this.setMatchReturnFlow(c.matchReturnFlow);
        if (c.inPortIds != null) {
            this.setInPorts(c.inPortIds.toArray(new UUID[c.inPortIds.size()]));
        }
        if (c.outPortIds != null) {
            this.setOutPorts(c.outPortIds.toArray(new UUID[c.outPortIds.size()]));
        }
        this.setPortGroup(c.portGroup);
        this.setIpAddrGroupDst(c.ipAddrGroupIdDst);
        this.setIpAddrGroupSrc(c.ipAddrGroupIdSrc);
        this.setDlType(c.etherType);
        if (null != c.ethSrc)
            this.setDlSrc(c.ethSrc.toString());
        if (NO_MASK != c.ethSrcMask)
            this.setDlSrcMask(MAC.maskToString(c.ethSrcMask));
        if (null != c.ethDst)
            this.setDlDst(c.ethDst.toString());
        if (NO_MASK != c.dlDstMask)
            this.setDlDstMask(MAC.maskToString(c.dlDstMask));
        if (null != c.nwDstIp) {
            this.setNwDstAddress(c.nwDstIp.getAddress().toString());
            this.setNwDstLength(c.nwDstIp.getPrefixLen());
        }
        if (null != c.nwSrcIp) {
            this.setNwSrcAddress(c.nwSrcIp.getAddress().toString());
            this.setNwSrcLength(c.nwSrcIp.getPrefixLen());
        }
        if (null != c.nwProto)
            this.setNwProto(unsign(c.nwProto));
        if (null != c.nwTos)
            this.setNwTos(unsign(c.nwTos));
        this.setFragmentPolicy(c.fragmentPolicy.toString().toLowerCase());
        if (null != c.tpDst)
            this.setTpDst(c.tpDst);
        if (null != c.tpSrc)
            this.setTpSrc(c.tpSrc);
    }

    protected boolean hasL4Fields() {
        return tpDst != null || tpSrc != null;
    }

    /**
     * If fragmentPolicy is null, returns the appropriate default
     * fragment policy based on the condition's other properties. If
     * not, verifies that the value of fragmentPolicy is consistent
     * with the condition's other properties and then returns it.
     *
     * @throws BadRequestHttpException if fragmentPolicy is non-null
     * and not permitted due to the condition's other properties.
     */
    protected FragmentPolicy getAndValidateFragmentPolicy() {
        if (getFragmentPolicy() == null)
            return hasL4Fields() ? FragmentPolicy.HEADER : FragmentPolicy.ANY;

        FragmentPolicy fp = FragmentPolicy.valueOf(fragmentPolicy.toUpperCase());
        if (hasL4Fields() && fp.accepts(IPFragmentType.Later))
            throw new BadRequestHttpException(MessageProperty.getMessage(
                    MessageProperty.FRAG_POLICY_INVALID_FOR_L4_RULE));

        return fp;
    }
}

File: python-midonetclient/src/midonetclient/condition.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2015 Midokura PTE LTD.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from midonetclient import resource_base
from midonetclient import vendor_media_type

class ConditionBase:
    def __init__(self, _dto):
        self._dto = _dto

    def is_cond_invert(self):
        return self._dto()['condInvert']

    def is_inv_dl_dst(self):
        return self._dto()['invDlDst']

    def is_inv_dl_src(self):
        return self._dto()['invDlSrc']

    def is_inv_dl_type(self):
        return self._dto()['invDlType']

    def is_inv_in_ports(self):
        return self._dto()['invInPorts']

    def is_inv_nw_dst(self):
        return self._dto()['invNwDst']

    def is_inv_nw_proto(self):
        return self._dto()['invNwProto']

    def is_inv_nw_src(self):
        return self._dto()['invNwSrc']

    def is_inv_nw_tos(self):
        return self._dto()['invNwTos']

    def is_inv_out_ports(self):
        return self._dto()['invOutPorts']

    def is_inv_port_group(self):
        return self._dto()['invPortGroup']

    def is_inv_ip_addr_group_dst(self):
        return self._dto()['invIpAddrGroupDst']

    def is_inv_ip_addr_group_src(self):
        return self._dto()['invIpAddrGroupSrc']

    def is_inv_tp_dst(self):
        return self._dto()['invTpDst']

    def is_inv_tp_src(self):
        return self._dto()['invTpSrc']

    def is_match_forward_flow(self):
        return self._dto()['matchForwardFlow']

    def is_match_return_flow(self):
        return self._dto()['matchReturnFlow']

    def get_chain_id(self):
        return self._dto()['chainId']

    def get_dl_dst(self):
        return self._dto()['dlDst']

    def get_dl_dst_mask(self):
        return self._dto()['dlDstMask']

    def get_dl_src(self):
        return self._dto()['dlSrc']

    def get_dl_src_mask(self):
        return self._dto()['dlSrcMask']

    def get_dl_type(self):
        return self._dto()['dlType']

    def get_flow_action(self):
        return self._dto()['flowAction']

    def get_id(self):
        return self._dto()['id']

    def get_in_ports(self):
        return self._dto()['inPorts']

    def get_nw_dst_address(self):
        return self._dto()['nwDstAddress']

    def get_nw_dst_length(self):
        return self._dto()['nwDstLength']

    def get_nw_proto(self):
        return self._dto()['nwProto']

    def get_nw_src_address(self):
        return self._dto()['nwSrcAddress']

    def get_nw_src_length(self):
        return self._dto()['nwSrcLength']

    def get_nw_tos(self):
        return self._dto()['nwTos']

    def get_out_ports(self):
        return self._dto()['outPorts']

    def get_port_group(self):
        return self._dto()['portGroup']

    def get_ip_addr_group_dst(self):
        return self._dto()['ipAddrGroupDst']

    def get_ip_addr_group_src(self):
        return self._dto()['ipAddrGroupSrc']

    def get_tp_src(self):
        return self._dto()['tpSrc']

    def get_tp_dst(self):
        return self._dto()['tpDst']

    def get_fragment_policy(self):
        return self._dto()['fragmentPolicy']

    def inv_port_group(self, inv_port_group):
        self._dto()['invPortGroup'] = inv_port_group
        return self

    def inv_ip_addr_group_dst(self, inv_ip_addr_group_dst):
        self._dto()['invIpAddrGroupDst'] = inv_ip_addr_group_dst
        return self

    def inv_ip_addr_group_src(self, inv_ip_addr_group_src):
        self._dto()['invIpAddrGroupSrc'] = inv_ip_addr_group_src
        return self

    def tp_src(self, tp_src):
        self._dto()['tpSrc'] = tp_src
        return self

    def dl_src(self, dl_src):
        self._dto()['dlSrc'] = dl_src
        return self

    def dl_src_mask(self, dl_src_mask):
        self._dto()['dlSrcMask'] = dl_src_mask
        return self

    def inv_nw_dst(self, inv_nw_dst):
        self._dto()['invNwDst'] = inv_nw_dst
        return self

    def dl_dst(self, dl_dst):
        self._dto()['dlDst'] = dl_dst
        return self

    def dl_dst_mask(self, dl_dst_mask):
        self._dto()['dlDstMask'] = dl_dst_mask
        return self

    def match_forward_flow(self, match_forward_flow):
        self._dto()['matchForwardFlow'] = match_forward_flow
        return self

    def inv_tp_src(self, inv_tp_src):
        self._dto()['invTpSrc'] = inv_tp_src
        return self

    def match_return_flow(self, match_return_flow):
        self._dto()['matchReturnFlow'] = match_return_flow
        return self

    def inv_nw_src(self, inv_nw_src):
        self._dto()['invNwSrc'] = inv_nw_src
        return self

    def out_ports(self, out_ports):
        self._dto()['outPorts'] = out_ports
        return self

    def nw_dst_length(self, nw_dst_length):
        self._dto()['nwDstLength'] = nw_dst_length
        return self

    def inv_out_ports(self, inv_out_ports):
        self._dto()['invOutPorts'] = inv_out_ports
        return self

    def dl_type(self, dl_type):
        self._dto()['dlType'] = dl_type
        return self

    def inv_nw_tos(self, inv_nw_tos):
        self._dto()['invNwTos'] = inv_nw_tos
        return self

    def port_group(self, port_group):
        self._dto()['portGroup'] = port_group
        return self

    def ip_addr_group_dst(self, ip_addr_group_dst):
        self._dto()['ipAddrGroupDst'] = ip_addr_group_dst
        return self

    def ip_addr_group_src(self, ip_addr_group_src):
        self._dto()['ipAddrGroupSrc'] = ip_addr_group_src
        return self

    def inv_dl_dst(self, inv_dl_dst):
        self._dto()['invDlDst'] = inv_dl_dst
        return self

    def inv_in_ports(self, inv_in_ports):
        self._dto()['invInPorts'] = inv_in_ports
        return self

    def inv_dl_type(self, inv_dl_type):
        self._dto()['invDlType'] = inv_dl_type
        return self

    def inv_tp_dst(self, inv_tp_dst):
        self._dto()['invTpDst'] = inv_tp_dst
        return self

    def chain_id(self, chain_id):
        self._dto()['chainId'] = chain_id
        return self

    def nw_tos(self, nw_tos):
        self._dto()['nwTos'] = nw_tos
        return self

    def nw_proto(self, nw_proto):
        self._dto()['nwProto'] = nw_proto
        return self

    def nw_src_length(self, nw_src_length):
        self._dto()['nwSrcLength'] = nw_src_length
        return self

    def in_ports(self, in_ports):
        self._dto()['inPorts'] = in_ports
        return self

    def nw_dst_address(self, nw_dst_address):
        self._dto()['nwDstAddress'] = nw_dst_address
        return self

    def nw_src_address(self, nw_src_address):
        self._dto()['nwSrcAddress'] = nw_src_address
        return self

    def inv_nw_proto(self, inv_nw_proto):
        self._dto()['invNwProto'] = inv_nw_proto
        return self

    def cond_invert(self, cond_invert):
        self._dto()['condInvert'] = cond_invert
        return self

    def tp_dst(self, tp_dst):
        self._dto()['tpDst'] = tp_dst
        return self

    def inv_dl_src(self, inv_dl_src):
        self._dto()['invDlSrc'] = inv_dl_src
        return self

    def fragment_policy(self, fragment_policy):
        self._dto()['fragmentPolicy'] = fragment_policy
        return self

    def is_no_vlan(self):
        return self._dto()['noVlan']

    def get_vlan(self):
        return self._dto()['vlan']

    def no_vlan(self, no_vlan):
        self._dto()['noVlan'] = no_vlan
        return self

    def vlan(self, vlan):
        self._dto()['vlan'] = vlan
        return self

class Condition(resource_base.ResourceBase, ConditionBase):

    media_type = vendor_media_type.APPLICATION_CONDITION_JSON

    def __init__(self, uri, dto, auth):
        super(Condition, self).__init__(uri, dto, auth)
        ConditionBase.__init__(self, lambda: dto)

